spacy/pipeline/coref.py

```diff
@@ -0,0 +1,330 @@
+                parent="CoreferenceResolver", method="add_label", name=self.name
+            )
+        )
+
+    def get_loss(
+        self,
+        examples: Iterable[Example],
+        score_matrix: Floats2d,
+        mention_idx: Ints2d,
+    ):
+        """Find the loss and gradient of loss for the batch of documents and
+        their predicted scores.
+
+        examples (Iterable[Examples]): The batch of examples.
+        scores: Scores representing the model's predictions.
+        RETURNS (Tuple[float, float]): The loss and the gradient.
+
+        DOCS: https://spacy.io/api/coref#get_loss (TODO)
+        """
+        ops = self.model.ops
+        xp = ops.xp
+
+        # TODO if there is more than one example, give an error
+        # (or actually rework this to take multiple things)
+        example = list(examples)[0]
+        cidx = mention_idx
+
+        clusters = get_clusters_from_doc(example.reference)
+        span_idxs = create_head_span_idxs(ops, len(example.predicted))
+        gscores = create_gold_scores(span_idxs, clusters)
+        # TODO fix type here. This is bools but asarray2f wants ints.
+        gscores = ops.asarray2f(gscores)
+        # top_gscores = xp.take_along_axis(gscores, cidx, axis=1)
+        top_gscores = xp.take_along_axis(gscores, mention_idx, axis=1)
+        # now add the placeholder
+        gold_placeholder = ~top_gscores.any(axis=1).T
+        gold_placeholder = xp.expand_dims(gold_placeholder, 1)
+        top_gscores = xp.concatenate((gold_placeholder, top_gscores), 1)
+
+        # boolean to float
+        top_gscores = ops.asarray2f(top_gscores)
+
+        with warnings.catch_warnings():
+            warnings.filterwarnings("ignore", category=RuntimeWarning)
+            log_marg = ops.softmax(score_matrix + ops.xp.log(top_gscores), axis=1)
+        log_norm = ops.softmax(score_matrix, axis=1)
+        grad = log_norm - log_marg
+        # gradients.append((grad, cidx))
+        loss = float((grad**2).sum())
+
+        return loss, grad
+
+    def initialize(
+        self,
+        get_examples: Callable[[], Iterable[Example]],
+        *,
+        nlp: Optional[Language] = None,
+    ) -> None:
+        """Initialize the pipe for training, using a representative set
+        of data examples.
+
+        get_examples (Callable[[], Iterable[Example]]): Function that
+            returns a representative sample of gold-standard Example objects.
+        nlp (Language): The current nlp object the component is part of.
+
+        DOCS: https://spacy.io/api/coref#initialize (TODO)
+        """
+        validate_get_examples(get_examples, "CoreferenceResolver.initialize")
+
+        X = []
+        Y = []
+        for ex in islice(get_examples(), 2):
+            X.append(ex.predicted)
+            Y.append(ex.reference)
+
+        assert len(X) > 0, Errors.E923.format(name=self.name)
+        self.model.initialize(X=X, Y=Y)
+
+    def score(self, examples, **kwargs):
```

Comment:
Also move this into a registered scorer with a scorer setting?

Reply:
Done in #10875.